### ISSUE TITLE
Init_net.pb & predict_net.pb, what's inside?

### ISSUE THREAD

**Issue by CriCL** [facebookarchive/caffe2#1850]
Hi everyone,

i'm new at caffe and would like to know what is inside init_net.pb.

i'm currently able to declare an object to read predict_net.pb, and show the "structure" of the model i have loaded. But, if i try to do the same to init_net.pb it crashes because of the size of the file.

If i want to "fine tune" a model, 
do i have to work(editing a layer) only with init_net.pb and write a new predict_net.pb file?

Hope someone can guide me,

Regards

**Comment by orionr** [comment]
Hi @CriCL. Welcome! The init_net typically stores all the weights, offsets, etc. used by the network, while the predict_net stores the operators that execute the net. Technically it's more like the init_net is just run once at the start and the predict_net is run repeatedly during inference / prediction.

To tune a model during training you're going to adjust the weights and offsets (what is in init_net) through something like SGD (stochastic gradient descent). You'd change predict_net when you're changing the model structure - moving from ResNet to ResNeXt or reducing the number of classes you'd want to predict (50 down to 16, for example).

More details at https://caffe2.ai/docs/intro-tutorial.html and https://caffe2.ai/docs/brew.html . Good luck!

**Comment by CriCL** [comment]
Hi @orionr , Thanks a lot for answering.

In fact, what i want to do is to retrieve top 10 similarities given an input image. This is the main reason for working with AlexNet and "retrain" layer fc7.

Given your explanation about init_net and predict_net, what i wonder now is:
do i have to modify predict_net.pb operators to make this work?

Somewhere in init_net i have to retrain the weights with my input images  and then use the predictor to retrieve similarities.

I really appreciate your collaboration in this project,

Regards

**Comment by orionr** [comment]
@CriCL if you want to retrain, you'll probably need to create a Python training script and use that to recreate the model. You can then load weights from the `pb` files and retrain there. We're looking to add training scripts to Caffe2 itself, but I'd recommend checking out the newly-released Detectron:

https://github.com/facebookresearch/Detectron

This provides all of the training data and scripts you need to create top-performing classification models like ResNet and ResNeXt and is based on Caffe2.

Alternatively, you can use, PyTorch to train a model like AlexNet and convert it to Caffe2 via ONNX:

http://pytorch.org/tutorials/advanced/super_resolution_with_caffe2.html

Lots of options out there!
